### ISSUE TITLE
HashLocationStrategy drops query params for default path

### ISSUE THREAD

**Issue by thomasturrell** [angular/angular#41265]
# 🐞 bug report

### Affected Package
 The issue is caused by HashLocationStrategy  in package @angular/common. Specifically the prepareExternalUrl function.

### Is this a regression?

Not that I am aware of.

### Description

HashLocationStrategy does not retain the query params in the URL when the application is initialised if the requested path is `''`.

For example the query params are dropped for the following:

- http://localhost:4200/?message=hello > http://localhost:4300/#/
- http://localhost:4200/index.html?message=hello > http://localhost:4300/#/

However the query params are retained for the following:

- localhost:4200/#/dashboard?message=hello

This is very surprising to me, I would like the query strings to be retained even when the application is requested without a path (or technically a fragment in this case).

Apologies if this is a feature request and not a bug. However it feels like a bug to me because the PathLocationStrategy does not behave in the same manner. You can verify this by turning off setting useHash to false in AppRoutingModule in my minimal reproduction.

## 🔬 Minimal Reproduction

You can see a minimal reproduction at [stackblitz](https://stackblitz.com/edit/angular-ivy-agczcd?file=src%2Fapp%2Fapp.component.html).

The top link drops the query params where as the bottom link retains the query params.

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 11.2.5
Node: 14.15.3
OS: darwin x64

Angular: 11.2.6
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1102.5
@angular-devkit/build-angular   0.1102.5
@angular-devkit/core            11.2.5
@angular-devkit/schematics      11.2.5
@angular/cdk                    11.2.5
@angular/cli                    11.2.5
@angular/material               11.2.5
@schematics/angular             11.2.5
@schematics/update              0.1102.5
ng-packagr                      11.2.4
rxjs                            6.6.6
typescript                      4.1.5

</code></pre>

**Anything else relevant?**

Issue #36688 provided me with a some ideas for a workaround. However issue #36688 required the query params to come before the fragments, I require the query params to be part of the fragment.

**Comment by thomasturrell** [comment]
This seems to be the source of the issue.

```typescript
  path(includeHash: boolean = false): string {
    // the hash value is always prefixed with a `#`
    // and if it is empty then it will stay empty
    let path = this._platformLocation.hash;
    if (path == null) path = '#';

    return path.length > 0 ? path.substring(1) : path;
  }
```

Doesn't this method always return ```#``` no matter what?

I am no expert so please forgive me if I am wrong.

**Comment by atscott** [comment]
Duplicate of #36688. Please refer to that discussion for details on how to resolve this for your application.

**Comment by thomasturrell** [comment]
> Duplicate of #36688. Please refer to that discussion for details on how to resolve this for your application.

@atscott Thank you for your response. I did refer to #36688. But it's unclear to me, why do HashLocationStrategy and PathLocationStrategy behave differently. PathLocationStrategy does not drop query params, where as HashLocationStrategy does.

> all URIs in the application have no path or query when using HashLocationStrategy - those are all moved to the fragment

This is not what I have observed, when opening the application the query parameters are dropped and not moved to the fragment. Is this desirable behaviour? It seems to be in contradiction of your statement.

Current behaviour:
http://localhost:4200/?message=hello is redirected to  http://localhost:4300/#/ 

I would have expected that
http://localhost:4200/?message=hello would be redirected to http://localhost:4300/#/?message=hello

Sorry about opening a new issue, I believed that I shouldn't comment on a closed issue.

**Comment by atscott** [comment]
> This is not what I have observed, when opening the application the query parameters are dropped and not moved to the fragment. Is this desirable behaviour?

Ah, this is not what that sentence means. It means that Angular interpreters query params by parsing them from the fragment. By using hash location strategy, you are telling Angular to use the fragment for navigations. It does not read the query params part (as defined by the spec) at all. Instead, it redefines the location of query params as being the after the first question mark in the fragment. Likewise, the “path” is not part is not what is defined on the spec, but rather part of the fragment as well. This is all expected and desired behavior.

To summarize, the path and query params parts as defined by the uri spec are not part of the Angular routing — these are all interpreted by parsing the fragment.
